## Das neue Flatfile-CSV-Format (ffcsv) der Datenbank GENESIS-Online

https://www-genesis.destatis.de/datenbank/beta/announcement  
  
Lauffähiger Python/Pandas Code zum Einlesen

In [1]:
import pandas as pd
pd.options.display.max_colwidth = 60

#### Das bleibt konstant:  
* Datensatz-orientiertes Format zur automatisierten Weiterverarbeitung
* Spalten durch Semikolon separiert (de, en)
* Dezimalkomma je nach Sprache "," (de) oder "." (en)
* utf-8 Codierung
* Zellen können wertersetzende Qualitätskennzeichen enthalten wie "-", "x", ".", "/"
* Spalte(n) Qualitätskennzeichen mit "_q" Suffix beim Download an- oder abwählbar

## Beispiel 1
`Tabelle: 61111-0001`  
**Verbraucherpreisindex: Deutschland, Jahre**  

#### Das bisherige ffcsv-Format  
* Preisindex und Veränderungsrate in getrennten Spalten
* Statistikspezifische Spaltenüberschriften

In [2]:
tabelle61111_0001alt = pd.read_csv("61111-0001_de_flat.csv", sep=";", decimal=",", 
                                   na_values=["-", "x", ".", "/"])

In [3]:
# ausgewählte Spalten
tabelle61111_0001alt[["Zeit","PREIS1__Verbraucherpreisindex__2020=100","PREIS1__Verbraucherpreisindex__q",
                   "Verbraucherpreisindex__CH0004","Verbraucherpreisindex__CH0004__q"]].tail(3)

Zeit  PREIS1__Verbraucherpreisindex__2020=100  \
30  2021                                    103.1   
31  2022                                    110.2   
32  2023                                    116.7   

   PREIS1__Verbraucherpreisindex__q  Verbraucherpreisindex__CH0004  \
30                                e                            3.1   
31                                e                            6.9   
32                                e                            5.9   

   Verbraucherpreisindex__CH0004__q  
30                                e  
31                                e  
32                                e

#### Das neue ffcsv-Format  
* Einheitliche englische Spaltenbeschriftungen für de/en 
* Eine Wertespalte "value" unabhängig von der Statistik
* Eine Spalte "value_unit" für die Einheit
* Auslieferung als ZIP-Datei (eine ffcsv-Datei je ZIP-Archiv)
* unsortiert

In [4]:
tabelle61111_0001neu = pd.read_csv("61111-0001_de_flat.zip", sep=";", 
                                   decimal=",", na_values=["-", "x", ".", "/"])\
                                .sort_values(by=["time","value_variable_label"])

In [5]:
# ausgewählte Spalten
tabelle61111_0001neu[["time","value","value_unit",
                      "value_variable_code","value_variable_label","value_q"]].tail(6)

time  value value_unit value_variable_code   value_variable_label value_q
47  2021  103.1   2020=100              PREIS1  Verbraucherpreisindex       e
46  2021    3.1          %              PREIS1                     in       e
45  2022  110.2   2020=100              PREIS1  Verbraucherpreisindex       e
44  2022    6.9          %              PREIS1                     in       e
41  2023  116.7   2020=100              PREIS1  Verbraucherpreisindex       e
40  2023    5.9          %              PREIS1                     in       e

**Tabellierung der Datensätze**

In [6]:
tabelle61111_0001neu.pivot(index="time", columns="value_unit", values="value").tail(3)

value_unit    %  2020=100
time                     
2021        3.1     103.1
2022        6.9     110.2
2023        5.9     116.7

## Beispiel 2
`Tabelle: 61111-0003`  
**Verbraucherpreisindex: Deutschland, Jahre, Klassifikation der Verwendungszwecke des Individualkonsums (COICOP 2-5-Steller Hierarchie)**  

#### Das bisherige ffcsv-Format  
* COICOP 2- bis 5-Steller gemischt in einer Spalte "2_Auspraegung_Code"
* 2- bis 5-Steller mussten über die Länge der Ausprägungscodes gefiltert werden

In [7]:
tabelle61111_0003alt = pd.read_csv("61111-0003_de_flat.csv", sep=";", decimal=",", 
                                   na_values=["-", "x", ".", "/"])

In [8]:
# ausgewählte Spalten
tabelle61111_0003alt[["Zeit","2_Merkmal_Code","2_Merkmal_Label",
                      "2_Auspraegung_Code", "2_Auspraegung_Label",
                      "PREIS1__Verbraucherpreisindex__2020=100"]].tail(15)

Zeit 2_Merkmal_Code                          2_Merkmal_Label  \
1910  2023         CC13A5  Verwendungszwecke des Individualkonsums   
1911  2023         CC13A5  Verwendungszwecke des Individualkonsums   
1912  2023         CC13A5  Verwendungszwecke des Individualkonsums   
1913  2023         CC13A5  Verwendungszwecke des Individualkonsums   
1914  2023         CC13A5  Verwendungszwecke des Individualkonsums   
1915  2023         CC13A5  Verwendungszwecke des Individualkonsums   
1916  2023         CC13A5  Verwendungszwecke des Individualkonsums   
1917  2023         CC13A5  Verwendungszwecke des Individualkonsums   
1918  2023         CC13A5  Verwendungszwecke des Individualkonsums   
1919  2023         CC13A5  Verwendungszwecke des Individualkonsums   
1920  2023         CC13A5  Verwendungszwecke des Individualkonsums   
1921  2023         CC13A5  Verwendungszwecke des Individualkonsums   
1922  2023         CC13A5  Verwendungszwecke des Individualkonsums   
1923  2023         CC13A5  Verwendungszwecke des Individualkonsums   
1924  2023         CC13A5  Verwendungszwecke des Individualkonsums   

     2_Auspraegung_Code  \
1910          CC13-1253   
1911         CC13-12532   
1912          CC13-1254   
1913         CC13-12541   
1914         CC13-12542   
1915          CC13-1255   
1916         CC13-12550   
1917          CC13-1262   
1918         CC13-12621   
1919         CC13-12622   
1920          CC13-1270   
1921         CC13-12701   
1922         CC13-12702   
1923         CC13-12703   
1924         CC13-12704   

                                           2_Auspraegung_Label  \
1910          Versicherungsdienstleistungen für die Gesundheit   
1911          Dienstl. priv. Unfall- und Krankenversicherungen   
1912             Versicherungsdienstleistungen für den Verkehr   
1913          Versicherungsdienstl. für private Verkehrsmittel   
1914        Versicherungsdienstleistungen für Reise und Gepäck   
1915                      Andere Versicherungsdienstleistungen   
1916                      Andere Versicherungsdienstleistungen   
1917                     Andere Finanzdienstleistungen, a.n.g.   
1918                Bank- und Sparkassengebühren und Ähnliches   
1919          Gebühren für Anlageberatung, Steuerberatung u.Ä.   
1920                           Andere Dienstleistungen, a.n.g.   
1921                                       Verwaltungsgebühren   
1922          Rechtsberatung, Rechtsanwalts- und Notargebühren   
1923                 Bestattungsleistungen und Friedhofsgebühr   
1924                      Andere Gebühren und Dienstleistungen   

      PREIS1__Verbraucherpreisindex__2020=100  
1910                                    107.0  
1911                                    107.0  
1912                                    110.6  
1913                                    110.6  
1914                                    108.8  
1915                                    106.4  
1916                                    106.4  
1917                                    109.8  
1918                                    109.8  
1919                                    109.9  
1920                                    113.2  
1921                                    111.2  
1922                                    114.4  
1923                                    113.2  
1924                                    109.1

**Für das Berichtsjahr 2023 enthält die bisherige ffcsv-Datei**
* 385 Ausprägungen zum Merkmal "Verwendungszwecke des Individualkonsums" (Merkmalscode = CC13A5)

In [9]:
tabelle61111_0003alt.query("Zeit==2023")["2_Merkmal_Code"].value_counts()

2_Merkmal_Code
CC13A5    385
Name: count, dtype: int64

#### Das neue ffcsv-Format
* Über die Spalte "2_variable_code" werden die verschiedenen COICOP-Hierarchien filterbar
* Jede COICOP-Hierarchiestufe ist separat vollständig auswertbar

In [10]:
tabelle61111_0003neu = pd.read_csv("61111-0003_de_flat.zip", sep=";", 
                                   decimal=",", na_values=["-", "x", ".", "/"])

In [11]:
tabelle61111_0003neu["2_variable_label"] = tabelle61111_0003neu["2_variable_label"].str.replace("-","‑") # non-breaking hyphen
tabelle61111_0003neu[tabelle61111_0003neu["2_variable_attribute_code"].str.contains("CC13-12")]\
                    [["time","2_variable_code", "2_variable_label",
                      "2_variable_attribute_code","2_variable_attribute_label","value","value_q"]]\
                    .sort_values(by=["time","2_variable_attribute_code"]).tail(15)                                                                     

time 2_variable_code  \
260   2023          CC13A4   
1383  2023          CC13A5   
1472  2023          CC13A5   
328   2023          CC13A4   
1650  2023          CC13A5   
639   2023          CC13A3   
489   2023          CC13A4   
1686  2023          CC13A5   
1658  2023          CC13A5   
745   2023          CC13A3   
162   2023          CC13A4   
865   2023          CC13A5   
1964  2023          CC13A5   
2063  2023          CC13A5   
1803  2023          CC13A5   

                                        2_variable_label  \
260   Verwendungszwecke des Individualkonsums, 4‑Steller   
1383  Verwendungszwecke des Individualkonsums, 5‑Steller   
1472  Verwendungszwecke des Individualkonsums, 5‑Steller   
328   Verwendungszwecke des Individualkonsums, 4‑Steller   
1650  Verwendungszwecke des Individualkonsums, 5‑Steller   
639   Verwendungszwecke des Individualkonsums, 3‑Steller   
489   Verwendungszwecke des Individualkonsums, 4‑Steller   
1686  Verwendungszwecke des Individualkonsums, 5‑Steller   
1658  Verwendungszwecke des Individualkonsums, 5‑Steller   
745   Verwendungszwecke des Individualkonsums, 3‑Steller   
162   Verwendungszwecke des Individualkonsums, 4‑Steller   
865   Verwendungszwecke des Individualkonsums, 5‑Steller   
1964  Verwendungszwecke des Individualkonsums, 5‑Steller   
2063  Verwendungszwecke des Individualkonsums, 5‑Steller   
1803  Verwendungszwecke des Individualkonsums, 5‑Steller   

     2_variable_attribute_code  \
260                  CC13-1254   
1383                CC13-12541   
1472                CC13-12542   
328                  CC13-1255   
1650                CC13-12550   
639                   CC13-126   
489                  CC13-1262   
1686                CC13-12621   
1658                CC13-12622   
745                   CC13-127   
162                  CC13-1270   
865                 CC13-12701   
1964                CC13-12702   
2063                CC13-12703   
1803                CC13-12704   

                              2_variable_attribute_label  value value_q  
260        Versicherungsdienstleistungen für den Verkehr  110.6       e  
1383    Versicherungsdienstl. für private Verkehrsmittel  110.6       e  
1472  Versicherungsdienstleistungen für Reise und Gepäck  108.8       e  
328                 Andere Versicherungsdienstleistungen  106.4       e  
1650                Andere Versicherungsdienstleistungen  106.4       e  
639                       Finanzdienstleistungen, a.n.g.  109.8       e  
489                Andere Finanzdienstleistungen, a.n.g.  109.8       e  
1686          Bank- und Sparkassengebühren und Ähnliches  109.8       e  
1658    Gebühren für Anlageberatung, Steuerberatung u.Ä.  109.9       e  
745                      Andere Dienstleistungen, a.n.g.  113.2       e  
162                      Andere Dienstleistungen, a.n.g.  113.2       e  
865                                  Verwaltungsgebühren  111.2       e  
1964    Rechtsberatung, Rechtsanwalts- und Notargebühren  114.4       e  
2063           Bestattungsleistungen und Friedhofsgebühr  113.2       e  
1803                Andere Gebühren und Dienstleistungen  109.1       e

**Für das Berichtsjahr 2023 enthält die neue ffcsv-Datei**
* 275 Ausprägungen zum Merkmal "Verwendungszwecke des Individualkonsums, 5-Steller" (Merkmalscode = CC13A5)
* 110 Ausprägungen zum Merkmal "Verwendungszwecke des Individualkonsums, 4-Steller" (Merkmalscode = CC13A4)
* 44 Ausprägungen zum Merkmal "Verwendungszwecke des Individualkonsums, 3-Steller" (Merkmalscode = CC13A3)
* 12 Ausprägungen zum Merkmal "Verwendungszwecke des Individualkonsums, 2-Steller" (Merkmalscode = CC13A2)

In [12]:
tabelle61111_0003neu.query("time==2023")["2_variable_code"].value_counts()

2_variable_code
CC13A5    275
CC13A4    110
CC13A3     44
CC13A2     12
Name: count, dtype: int64

**Tabellierung der Datensätze**

In [13]:
tabelle61111_0003neu.pivot(index=["2_variable_label","2_variable_attribute_label"], 
                                    columns="time", values="value")

time                                                                                                   2019  \
2_variable_label                                   2_variable_attribute_label                                 
Verwendungszwecke des Individualkonsums, 2‑Steller Alkoholische Getränke und Tabakwaren                98.1   
                                                   Andere Waren und Dienstleistungen                   97.5   
                                                   Bekleidung und Schuhe                              101.7   
                                                   Bildungswesen                                      100.2   
                                                   Freizeit, Unterhaltung und Kultur                  100.3   
...                                                                                                     ...   
Verwendungszwecke des Individualkonsums, 5‑Steller Ärztliche Dienstleistungen der Allgemeinmedizin    100.0   
                                                   Ärztliche Dienstleistungen der Fachmedizin         100.0   
                                                   Übernachtungen in Feriendörfern, Camping u.Ä.       97.1   
                                                   Übernachtungen in Hotels, Gasthöfen und Ähnlichem   99.7   
                                                   Übernachtungen in anderen Unterkünften              98.3   

time                                                                                                   2020  \
2_variable_label                                   2_variable_attribute_label                                 
Verwendungszwecke des Individualkonsums, 2‑Steller Alkoholische Getränke und Tabakwaren               100.0   
                                                   Andere Waren und Dienstleistungen                  100.0   
                                                   Bekleidung und Schuhe                              100.0   
                                                   Bildungswesen                                      100.0   
                                                   Freizeit, Unterhaltung und Kultur                  100.0   
...                                                                                                     ...   
Verwendungszwecke des Individualkonsums, 5‑Steller Ärztliche Dienstleistungen der Allgemeinmedizin    100.0   
                                                   Ärztliche Dienstleistungen der Fachmedizin         100.0   
                                                   Übernachtungen in Feriendörfern, Camping u.Ä.      100.0   
                                                   Übernachtungen in Hotels, Gasthöfen und Ähnlichem  100.0   
                                                   Übernachtungen in anderen Unterkünften             100.0   

time                                                                                                   2021  \
2_variable_label                                   2_variable_attribute_label                                 
Verwendungszwecke des Individualkonsums, 2‑Steller Alkoholische Getränke und Tabakwaren               103.5   
                                                   Andere Waren und Dienstleistungen                  103.8   
                                                   Bekleidung und Schuhe                              101.5   
                                                   Bildungswesen                                      102.5   
                                                   Freizeit, Unterhaltung und Kultur                  102.9   
...                                                                                                     ...   
Verwendungszwecke des Individualkonsums, 5‑Steller Ärztliche Dienstleistungen der Allgemeinmedizin    100.0   
                                                   Ärztliche Dienstleistungen der Fachmedizin         100.0   
  

## Beispiel 3
`Tabelle: 21611-0002`  
**Kinos, Leinwände, Sitzplätze der Kinos, Filmbesuche, Durchschnittlicher Kino-Eintrittspreis, Einnahmen, Filmabgabe: Deutschland, Jahre**  

#### Das bisherige ffcsv-Format  
* Statistik- und sprachspezifische (de/en) Spaltenüberschriften, die sich aus Code, Label und Einheit zusammensetzen

In [14]:
tabelle21611_0002alt = pd.read_csv("21611-0002_de_flat.csv", sep=";", decimal=",", 
                                   na_values=["-", "x", ".", "/"])

In [15]:
# ausgewählte Spalten
tabelle21611_0002alt[["Zeit","FILM11__Kinos__Anzahl","FILM08__Sitzplaetze_der_Kinos__EUR",
                      "FILM05__Filmbesuche__Mill._EUR","FILM10__Filmbesuche_je_Einwohner__Mill._EUR"]].tail(2)

Zeit  FILM11__Kinos__Anzahl  FILM08__Sitzplaetze_der_Kinos__EUR  \
21  2021                   1723                                8.87   
22  2022                   1730                                9.26   

    FILM05__Filmbesuche__Mill._EUR  \
21                           373.2   
22                           722.0   

    FILM10__Filmbesuche_je_Einwohner__Mill._EUR  
21                                        369.2  
22                                        715.0

#### Das neue ffcsv-Format  
* Getrennte Spalten für Code, Label und Einheit
* Programme statistikübergreifend nutzbar

In [16]:
tabelle21611_0002neu = pd.read_csv("21611-0002_de_flat.zip", sep=";", decimal=",", 
                                   na_values=["-", "x", ".", "/"])

In [17]:
# ausgewählte Spalten
tabelle21611_0002neu[["time","value","value_unit","value_variable_code",
                      "value_variable_label","value_q"]]\
                    .sort_values(by=["time","value_variable_code"]).tail(18)

time      value value_unit value_variable_code  \
126  2021    4931.00     Anzahl              FILM02   
132  2021  789688.00     Anzahl              FILM03   
128  2021      42.10      Mill.              FILM04   
127  2021     373.20  Mill. EUR              FILM05   
131  2021       0.50     Anzahl              FILM07   
129  2021       8.87        EUR              FILM08   
133  2021       4.00  Mill. EUR              FILM09   
134  2021     369.20  Mill. EUR              FILM10   
130  2021    1723.00     Anzahl              FILM11   
108  2022    4911.00     Anzahl              FILM02   
114  2022  777844.00     Anzahl              FILM03   
110  2022      78.00      Mill.              FILM04   
109  2022     722.00  Mill. EUR              FILM05   
113  2022       0.90     Anzahl              FILM07   
111  2022       9.26        EUR              FILM08   
115  2022       7.00  Mill. EUR              FILM09   
116  2022     715.00  Mill. EUR              FILM10   
112  2022    1730.00     Anzahl              FILM11   

                                   value_variable_label value_q  
126                                           Leinwände       e  
132                                Sitzplätze der Kinos       e  
128                                         Filmbesuche       e  
127                  Bruttoeinnahmen aus dem Filmbesuch       e  
131                            Filmbesuche je Einwohner       e  
129              Durchschnittlicher Kino-Eintrittspreis       e  
133                     Filmabgabe der Kinos an die FFA       e  
134  Nettoeinnahmen aus dem Filmbesuch (einschl. MwSt.)       e  
130                                               Kinos       e  
108                                           Leinwände       e  
114                                Sitzplätze der Kinos       e  
110                                         Filmbesuche       e  
109                  Bruttoeinnahmen aus dem Filmbesuch       e  
113                            Filmbesuche je Einwohner       e  
111              Durchschnittlicher Kino-Eintrittspreis       e  
115                     Filmabgabe der Kinos an die FFA       e  
116  Nettoeinnahmen aus dem Filmbesuch (einschl. MwSt.)       e  
112                                               Kinos       e

**Tabellierung der Datensätze**

In [18]:
tabelle21611_0002neu.query("time>2020").pivot(index=["value_variable_label","value_unit"], 
                                              columns="time", values="value")

time                                                                2021  \
value_variable_label                               value_unit              
Bruttoeinnahmen aus dem Filmbesuch                 Mill. EUR      373.20   
Durchschnittlicher Kino-Eintrittspreis             EUR              8.87   
Filmabgabe der Kinos an die FFA                    Mill. EUR        4.00   
Filmbesuche                                        Mill.           42.10   
Filmbesuche je Einwohner                           Anzahl           0.50   
Kinos                                              Anzahl        1723.00   
Leinwände                                          Anzahl        4931.00   
Nettoeinnahmen aus dem Filmbesuch (einschl. MwSt.) Mill. EUR      369.20   
Sitzplätze der Kinos                               Anzahl      789688.00   

time                                                                2022  
value_variable_label                               value_unit             
Bruttoeinnahmen aus dem Filmbesuch                 Mill. EUR      722.00  
Durchschnittlicher Kino-Eintrittspreis             EUR              9.26  
Filmabgabe der Kinos an die FFA                    Mill. EUR        7.00  
Filmbesuche                                        Mill.           78.00  
Filmbesuche je Einwohner                           Anzahl           0.90  
Kinos                                              Anzahl        1730.00  
Leinwände                                          Anzahl        4911.00  
Nettoeinnahmen aus dem Filmbesuch (einschl. MwSt.) Mill. EUR      715.00  
Sitzplätze der Kinos                               Anzahl      777844.00